Semantics.md

```diff
@@ -678,6 +829,51 @@ outside the range which rounds to an integer in range) traps.
     This trap is intended to be impossible for user code to catch or handle, even in the future when it may be possible to
     handle some other kinds of traps or exceptions.
 
+## Thread operators
+
+  * `is_lock_free`: Given an operand `N`, if the atomic step of an atomic
+    primitive (see [Atomic Memory Accesses](#atomic-memory-accesses)) on a
+    datum of size `N` bytes will be performed without the agent acquiring a
+    lock outside the `N` bytes comprising the datum, then return `1`.
+    Otherwise, return `0`.
+
+### Wait
+
+The wake and wait operators are optimizations over busy-waiting for a value to
+change. It is a validation error to use these operators on non-shared linear
+memory.
+
+The wait operators take three operands: an address operand, an expected
+value, and a relative timeout in milliseconds as an f64. The timeout must be
+non-negative, and can also be positive infinity. If the timeout is positive
+infinity then the timeout will never expire.
+
+The wait operation begins by performing an atomic load from the given address.
+If the loaded value is not equal to the expected value, the operator returns 1
+("not-equal"). If the values are equal, the thread is suspended. If the thread
+is woken (by execution of the corresponding "wake" operator on another thread),
+the wait operator returns 0 ("ok"). If the timeout expires before another
+thread wakes this one, this operator returns 2 ("timed-out").
+
+| Value | Description |
+| ---- | ---- |
+| 0 | "ok", woken by another thread |
+| 1 | "not-equal", the loaded value did not match the expected value |
+| 2 | "timed-out", not woken by another thread before timeout expired |
+
+  * `i32.wait`: load i32 value, compare to expected, and wait for `i32.wake` at same address
+  * `i64.wait`: load i64 value, compare to expected, and wait for `i64.wake` at same address
+
+### Wake
+
+The wake operators take two operands: an address operand, and a wake count. The
+operation will wake a maximum of "wake count" threads that are waiting on the
+same address. The operator returns the number of thread that were
+woken.
+
+  * `i32.wake`: wake up to N threads waiting on the given address via `i32.wait`
+  * `i64.wake`: wake up to N threads waiting on the given address via `i64.wait`
```

Comment:
Yeah, this is a little clunky. I made it a signed `i32`, where < 0 means wake all. Using an `f64` for the wake count seemed really odd. Perhaps it could be a separate opcode?